cmake/Modules/FindNCCL.cmake

```diff
@@ -33,6 +33,23 @@
 
 set(NCCL_ROOT_DIR "" CACHE PATH "Folder contains NVIDIA NCCL")
 
+# first check in the /usr/local/cuda before other paths
```

Comment:
"first check" may break the assumption that `NCCL_ROOT_DIR` can be used to specify nccl root directory?

Would adding `/usr/local/cuda` as last item in `find_path`, `find_library` below be sufficient? Or does it throw an error if `/usr/local/cuda` doesn't exist?

Reply:
I'm not sure why it breaks the assumption. Can you help me rephrase the comment? I can remove "first" if that's what confuses the users.

Reply:
Quoting from https://cmake.org/cmake/help/latest/command/find_path.html

> This command is used to find a directory containing the named file. A cache entry named by `<VAR>` is created to store the result of this command. If the file in a directory is found the result is stored in the variable and the search will not be repeated unless the variable is cleared. If nothing is found, the result will be `<VAR>-NOTFOUND`, and the search will be attempted again the next time find_path is invoked with the same variable.

Thus if there is some `nccl.h` at `/usr/local/cuda` but the user specifies `NCCL_ROOT_DIR=/home/ubuntu/myversionofnccl`, the `/usr/local/cuda` is wrongly used? Or do I misunderstand the doc? 

Btw, as per https://cmake.org/cmake/help/latest/policy/CMP0074.html `NCCL_ROOT_DIR` should probably be called `NCCL_ROOT` (or `NCCL_ROOT` should be added in addition to `NCCL_ROOT_DIR`).

Reply:
Ya. Even I saw this NCCL_ROOT and found that NCCL_ROOT_DIR is deprecated hence no longer to be used in PyTorch repo
https://github.com/pytorch/pytorch/blob/f94aab45fd7b1e7e32d8ea4324da6d3632a0d412/cmake/Modules/FindNCCL.cmake#L21

However, I couldn't find official cmake doc saying so. Nonetheless, I think I should add that too.

Reply:
> If there is some nccl.h at /usr/local/cuda but the user specifies NCCL_ROOT_DIR=/home/ubuntu/myversionofnccl, the /usr/local/cuda is wrongly used? Or do I misunderstand the doc?

I agree, since "search" will not be repeated until var is set, it makes sense to assume whatever location is found first is final. So in that case first check has to be env var and second for default.

Reply:
Surprisingly, I couldn't find a check for `/usr/local/cuda` in case of pytorch's FindNCCL
I have created a discuss question on their forum anyway - https://discuss.pytorch.org/t/findnccl-cmake-search-for-usr-local-cuda/66891

Reply:
> I saw this NCCL_ROOT and found that NCCL_ROOT_DIR is deprecated hence no longer to be used in PyTorch repo
> https://github.com/pytorch/pytorch/blob/f94aab45fd7b1e7e32d8ea4324da6d3632a0d412/cmake/Modules/FindNCCL.cmake#L21
> 
> However, I couldn't find official cmake doc saying so. Nonetheless, I think I should add that too.

Thanks. The official doc should be https://cmake.org/cmake/help/latest/policy/CMP0074.html

Reply:
Moved /usr/local/cuda check to be after other checks
Also added NCCL_ROOT alongside NCCL_ROOT_DIR
Also added deprecation warning for ROOT_DIR

Reply:
@ChaiBapchya Thank you!

After reading CMP0074 again,
> [...] now searches prefixes specified by the `<PackageName>_ROOT` CMake variable and the `<PackageName>_ROOT` environment variable. Package roots are maintained as a stack so nested calls to all find_* commands inside find modules and config packages also search the roots as prefixes.

seems to imply that it's not necessary to list `NCCL_ROOT` as part of `find_path` and `find_library`, but rather it's done automatically. Do you understand it in the same way?
Also https://cmake.org/cmake/help/latest/command/find_package.html#search-procedure (paragraph "The set of installation prefixes is constructed using the following steps [...]") suggests this.

If so,  `set(NCCL_ROOT $ENV{NCCL_ROOT} CACHE PATH "Folder contains NVIDIA NCCL")` would be superfluous. Also it seems none of the other Find modules check the `X_LIB_DIR` and `X_LIB_DIR` env variables.

Reply:
1. NCCL_ROOT auto check - I guess it is done automatically. Though it's not mentioned explicitly anywhere. Hence asked here - https://stackoverflow.com/questions/59783425/does-cmake-find-path-find-library-search-for-packagename-root-by-default

2. 
> Also it seems none of the other Find modules check the X_LIB_DIR and X_LIB_DIR env variables.
Not sure what you implying? You think that is superfluous too?

Reply:
1. Autocheck will be done provided `NO_DEFAULT_PATH` or `NO_PACKAGE_ROOT_PATH` are not set. So I'm assuming that's the case.
Refer : https://stackoverflow.com/questions/59783425/does-cmake-find-path-find-library-search-for-packagename-root-by-default/59788313#59788313

Reply:
1. Yes, you should remove the manual checks for `NCCL_ROOT`.

2. I don't think it's superfluous (ie it does something), but it seems not necessary because it seems non-standard (ie other Find modules don't check the environment variables, so it's not expected that you check them here)